### PR TITLE
add changes for enabling AML run

### PR DESCRIPTION
**Pull request by msp8955**
Adding an example script that can be used to run Megatron-DS using Azure ML and any necessary changes to enable it.

### CODE REVIEW

File: examples/azureml/Dockerfile.dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM mcr.microsoft.com/azureml/aifx/stable-ubuntu2004-cu115-py38-torch1110
+
+USER root:root
+
+RUN pip install pybind11
+
+RUN pip install git+https://github.com/microsoft/DeepSpeed.git@azure
```
Comment (awan-10) [inline examples/azureml/Dockerfile.dockerfile:7]: @msp8955 -- you can use the public DeepSpeed master branch and no longer need the azure branch.

File: megatron/initialize.py [hunk 2]
```diff
@@ -274,3 +274,15 @@ def _initialize_mem_buffs():
     # Initialize memory for checkpointed activations.
     if args.distribute_checkpointed_activations:
         mpu.init_checkpointed_activations_memory_buffer()
+
+def _is_rank_0():
+    """Check whether it is rank 0. For AML, check if it is rank 0 of a node"""
+    if torch.distributed.is_initialized():
+        if torch.distributed.get_rank() == 0 or (
+            'AZUREML_EXPERIMENT_ID' in os.environ and torch.distributed.get_rank() % torch.cuda.device_count() == 0
```
Comment (awan-10) [inline megatron/initialize.py:282]: can we use is_aml() function here to simplify the if check

**Comment by awan-10** [comment]
@jeffra -- please review this PR. Happy to discuss this over a quick call. There are some rough ends but so I gave some comments to @msp8955. For now, I feel that this PR is mergeable and we can do follow-up cleaning afterward.

File: megatron/initialize.py (hunk 2 above)
Comment (msp8955) [reply megatron/initialize.py:282]: __init__.py requires this initialize.py so it's having circular dependencies if I try to use it. Is there another place I can put those?

File: examples/azureml/aml_submit.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+megatron_ds_env.environment_variables['NCCL_SOCKET_IFNAME'] = 'eth0'
+megatron_ds_env.environment_variables['NCCL_IB_PCI_RELAXED_ORDERING']='1'
+megatron_ds_env.environment_variables['UCX_TLS']='tcp'
+megatron_ds_env.environment_variables['UCX_NET_DEVICES']='eth0'
+
+#-------------------------------------------------------------------------------
+# Training Settings and Arguments
+#-------------------------------------------------------------------------------
+node_count = 2
+total_proceses_count = 16
```
Comment (awan-10) [inline examples/azureml/aml_submit.py:81]: spelling mistake to fix: total_proceses_count --> total_processes or total_process_count

File: examples/azureml/aml_submit.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            '--adam-beta2', 0.95,
+            '--init-method-std', 0.006,
+            '--fp16',
+            '--data-impl', 'mmap',
+            '--checkpoint-activations',
+            '--tensorboard-dir', tensorboard_dir,
+            #'--cpu-optimizer',
+            '--deepspeed',
+            '--no-pipeline-parallel',
+            '--deepspeed_config', 'ds_config.json',
```
[the remark this answers is not in the document]
Comment (awan-10) [reply examples/azureml/aml_submit.py:121]: Thanks for doing this change. It looks good.

**Comment by awan-10** [comment]
@msp8955 -- all good here. Please fix the spelling problem and we are good to merge this PR.

**Pull request merged**